Use domain name for outbound TLS connections

**Is your feature request related to a problem? Please describe.**

When Haraka tries to connect to another host (outbound), it uses the IP for SNI and to validate the certificate.

**Describe the solution you'd like**

Haraka sends the domain as SNI and also validates the certificate using the domain. I would argue that nowadays, certificates usually contain a domain instead of a static IP.

**Describe alternatives you've considered**

Having certificates with static IPs is not really an alternative for us. Of course, sending the domain instead of the IP might also be configurable in the config.

**Additional context**

Maybe, I am just using Haraka wrongly, so here is some information on my setup:

Haraka | Haraka.js — Version: 3.0.1
 --- | :---
Node | v21.6.1
OS | Linux example.fauringer.de 6.7.3-arch1-1 #1 SMP PREEMPT_DYNAMIC Thu, 01 Feb 2024 10:30:35 +0000 x86_64 GNU/Linux
openssl | OpenSSL 1.1.1w  11 Sep 2023

Here are also some logs (note the node deprecation warning):

```
[INFO] [-] [core] [outbound] Transaction delivery for domain: example.fauringer.de
[INFO] [-] [core] loading tls.ini
[DEBUG] [4DCF459F-A46F-4010-A8C2-37011FC70164.1.1] [outbound] running send_email hooks
[DEBUG] [4DCF459F-A46F-4010-A8C2-37011FC70164.1.1] [outbound] Sending mail: 1707513574559_1707513574559_0_28_4DtbPH_1_example.fauringer.de
[DEBUG] [4DCF459F-A46F-4010-A8C2-37011FC70164.1.1] [outbound] running get_mx hooks
[DEBUG] [4DCF459F-A46F-4010-A8C2-37011FC70164.1.1] [outbound] running get_mx hook in queue/lmtp plugin
[DEBUG] [4DCF459F-A46F-4010-A8C2-37011FC70164.1.1] [outbound]  hook=get_mx plugin=queue/lmtp function=hook_get_mx params=example.fauringer.de retval=CONT msg=""
[DEBUG] [4DCF459F-A46F-4010-A8C2-37011FC70164.1.1] [outbound] delivering from: example.fauringer.de to: 192.168.200.4:25 (0) (0)
[DEBUG] [75D90002-135C-4EBC-9092-9EB13C1842B4] [core] [outbound] created. host: 192.168.200.4 port: 25
[NOTICE] [CDB9268D-5A4C-4033-84DB-31BDFA292614] [core] connect ip=192.168.200.4 port=39864 local_ip=192.168.200.4 local_port=25
[DEBUG] [CDB9268D-5A4C-4033-84DB-31BDFA292614] [core] running connect_init hooks
[DEBUG] [CDB9268D-5A4C-4033-84DB-31BDFA292614] [core] running connect_init hook in early_talker plugin
[DEBUG] [CDB9268D-5A4C-4033-84DB-31BDFA292614] [core]  hook=connect_init plugin=early_talker function=early_talker params="" retval=CONT msg=""
[DEBUG] [CDB9268D-5A4C-4033-84DB-31BDFA292614] [core] running connect_init_respond
[DEBUG] [CDB9268D-5A4C-4033-84DB-31BDFA292614] [core] running lookup_rdns hooks
[DEBUG] [CDB9268D-5A4C-4033-84DB-31BDFA292614] [core] running connect hooks
[DEBUG] [CDB9268D-5A4C-4033-84DB-31BDFA292614] [core] running ehlo hooks
[DEBUG] [CDB9268D-5A4C-4033-84DB-31BDFA292614] [core] running ehlo hook in helo.checks plugin
[DEBUG] [CDB9268D-5A4C-4033-84DB-31BDFA292614] [core]  hook=ehlo plugin=helo.checks function=proto_mismatch_esmtp params=example.fauringer.de retval=CONT msg=""
[DEBUG] [CDB9268D-5A4C-4033-84DB-31BDFA292614] [core] running ehlo hook in helo.checks plugin
[DEBUG] [CDB9268D-5A4C-4033-84DB-31BDFA292614] [core]  hook=ehlo plugin=helo.checks function=init params=example.fauringer.de retval=CONT msg=""
[DEBUG] [CDB9268D-5A4C-4033-84DB-31BDFA292614] [core] running ehlo hook in helo.checks plugin
[INFO] [CDB9268D-5A4C-4033-84DB-31BDFA292614] [helo.checks] skip:proto_mismatch(private)
[DEBUG] [CDB9268D-5A4C-4033-84DB-31BDFA292614] [core]  hook=ehlo plugin=helo.checks function=emit_log params=example.fauringer.de retval=CONT msg=""
[DEBUG] [CDB9268D-5A4C-4033-84DB-31BDFA292614] [core] running capabilities hooks
[DEBUG] [CDB9268D-5A4C-4033-84DB-31BDFA292614] [core] running capabilities hook in tls plugin
[DEBUG] [CDB9268D-5A4C-4033-84DB-31BDFA292614] [core]  hook=capabilities plugin=tls function=advertise_starttls params="" retval=CONT msg=""
[DEBUG] [4DCF459F-A46F-4010-A8C2-37011FC70164.1.1] [outbound] Trying TLS for domain: example.fauringer.de, host: 192.168.200.4
[DEBUG] [CDB9268D-5A4C-4033-84DB-31BDFA292614] [core] running unrecognized_command hooks
[DEBUG] [CDB9268D-5A4C-4033-84DB-31BDFA292614] [core] running unrecognized_command hook in tls plugin
[DEBUG] [-] [core] Upgrading to TLS
[DEBUG] [-] [core] client TLS upgrade in progress, awaiting secured.
(node:28) [DEP0123] DeprecationWarning: Setting the TLS ServerName to an IP address is not permitted by RFC 6066. This will be ignored in a future version.
[DEBUG] [-] [core] SNI servername: 192.168.200.4
[ERROR] [-] [core] client TLS error: Error [ERR_TLS_CERT_ALTNAME_INVALID]: Hostname/IP does not match certificate's altnames: IP: 192.168.200.4 is not in the cert's list:
[ERROR] [4DCF459F-A46F-4010-A8C2-37011FC70164.1.1] [outbound] Ongoing connection failed to 192.168.200.4:25 : Error [ERR_TLS_CERT_ALTNAME_INVALID]: Hostname/IP does not match certificate's altnames: IP: 192.168.200.4 is not in the cert's list:
[DEBUG] [-] [core] [outbound] release_client: 75D90002-135C-4EBC-9092-9EB13C1842B4 192.168.200.4:25 to undefined
[DEBUG] [-] [core] Temp fail for: Tried all MXs
```

--- a/Changes.md
+++ b/Changes.md
@@ -1,9 +1,15 @@
 
 ### Unreleased
 
+#### Changed
+
+- outbound/mx_lookup: make it async/await
+
 #### Fixed
 
 - fix(outbound): allow LHLO over insecure socket if TLS is forced #3278
+- outbound: emit log message when ignoring local MX
+- fix(outbound): don't send SNI servername when connecting to an IP
 
 ### [3.0.3] - 2024-02-07
 

--- a/outbound/hmail.js
+++ b/outbound/hmail.js
@@ -227,9 +227,7 @@ class HMailItem extends events.EventEmitter {
         }
 
         // if none of the above return codes, drop through to this...
-        mx_lookup.lookup_mx(this.todo.domain, (err, mxs) => {
-            this.found_mx(err, mxs);
-        });
+        mx_lookup.lookup_mx(this.todo.domain, this.found_mx)
     }
 
     found_mx (err, mxs) {
@@ -287,7 +285,7 @@ class HMailItem extends events.EventEmitter {
         }
     }
 
-    try_deliver () {
+    async try_deliver () {
 
         // check if there are any MXs left
         if (this.mxlist.length === 0) {
@@ -299,6 +297,13 @@ class HMailItem extends events.EventEmitter {
 
         const mx = this.mxlist.shift();
         const host = mx.exchange;
+
+        if (!obc.cfg.local_mx_ok) {
+            if (await net_utils.is_local_host(host)) {
+                this.loginfo(`MX ${host} is local, skipping since local_mx_ok=false`)
+                return this.try_deliver(); // try next MX
+            }
+        }
 
         this.force_tls = this.todo.force_tls;
         if (!this.force_tls) {

--- a/outbound/mx_lookup.js
+++ b/outbound/mx_lookup.js
@@ -1,11 +1,9 @@
 "use strict";
 
-const dns         = require('dns');
-const net_utils   = require('haraka-net-utils')
+const dns        = require('node:dns').promises;
+const net_utils  = require('haraka-net-utils')
 
-const obc         = require('./config');
-
-exports.lookup_mx = function lookup_mx (domain, cb) {
+exports.lookup_mx = async function lookup_mx (domain, cb) {
     const mxs = [];
 
     // Possible DNS errors
@@ -22,49 +20,41 @@ exports.lookup_mx = function lookup_mx (domain, cb) {
     // EREFUSED
     // SERVFAIL
 
-    // default wrap_mx just returns our object with "priority" and "exchange" keys
-    let wrap_mx = a => a;
-    async function process_dns (err, addresses) {
-        if (err) {
-            if (err.code === 'ENODATA' || err.code === 'ENOTFOUND') {
-                // Most likely this is a hostname with no MX record
-                // Drop through and we'll get the A record instead.
-                return 0;
-            }
-            cb(err);
+    try {
+        const addresses = await net_utils.get_mx(domain)
+        for (const a of addresses) {
+            mxs.push(a);
         }
-        else if (addresses?.length) {
-            for (let i=0,l=addresses.length; i < l; i++) {
-                if (
-                    obc.cfg.local_mx_ok ||
-                    await net_utils.is_local_host(addresses[i].exchange).catch(() => null) === false
-                ) {
-                    const mx = wrap_mx(addresses[i]);
-                    mxs.push(mx);
-                }
-            }
-            cb(null, mxs);
+        if (mxs.length) {
+            if (cb) return cb(null, mxs)
+            return mxs
         }
-        else {
-            // return zero if we need to keep trying next option
-            return 0;
+    }
+    catch (err) {
+        switch (err.code) {
+          case 'ENODATA':
+          case 'ENOTFOUND':
+            // likely a hostname with no MX record, drop through
+            break
+          default:
+            throw err(err)
         }
-        return 1;
     }
 
-    net_utils.get_mx(domain, async (err, addresses) => {
-        if (await process_dns(err, addresses)) return;
+    // No MX record, try resolving A record
 
-        // if MX lookup failed, we lookup an A record. To do that we change
-        // wrap_mx() to return same thing as resolveMx() does.
-        wrap_mx = a => ({priority:0,exchange:a});
-        // IS: IPv6 compatible
-        dns.resolve(domain, async (err2, addresses2) => {
-            if (await process_dns(err2, addresses2)) return;
+    // wrap addresses with "priority" and "exchange" keys
+    let wrap_mx = a => ({priority:0,exchange:a});
 
-            err2 = new Error("Found nowhere to deliver to");
-            err2.code = 'NOMX';
-            cb(err2);
-        });
-    });
+    const addresses = await dns.resolve(domain)
+    for (const a of addresses) {
+        mxs.push(wrap_mx(a));
+    }
+
+    if (mxs.length) return mxs
+
+    const err = new Error("Found nowhere to deliver to");
+    err.code = 'NOMX';
+    if (cb) return cb(err)
+    throw err
 }

--- a/outbound/mx_lookup.js
+++ b/outbound/mx_lookup.js
@@ -32,19 +32,19 @@ exports.lookup_mx = async function lookup_mx (domain, cb) {
     }
     catch (err) {
         switch (err.code) {
-          case 'ENODATA':
-          case 'ENOTFOUND':
-            // likely a hostname with no MX record, drop through
-            break
-          default:
-            throw err(err)
+            case 'ENODATA':
+            case 'ENOTFOUND':
+                // likely a hostname with no MX record, drop through
+                break
+            default:
+                throw err(err)
         }
     }
 
     // No MX record, try resolving A record
 
     // wrap addresses with "priority" and "exchange" keys
-    let wrap_mx = a => ({priority:0,exchange:a});
+    const wrap_mx = a => ({priority:0,exchange:a});
 
     const addresses = await dns.resolve(domain)
     for (const a of addresses) {

--- a/outbound/tls.js
+++ b/outbound/tls.js
@@ -1,5 +1,7 @@
 'use strict';
 
+const net          = require('net')
+
 const logger       = require('../logger');
 const tls_socket   = require('../tls_socket');
 const config       = require('haraka-config');
@@ -69,6 +71,7 @@ class OutboundTLS {
     }
 
     get_tls_options (mx) {
+        if (net.isIP(mx.exchange)) return this.cfg
         return Object.assign(this.cfg, {servername: mx.exchange});
     }
 

--- a/package.json
+++ b/package.json
@@ -17,11 +17,11 @@
   },
   "main": "haraka.js",
   "engines": {
-    "node": ">=16"
+    "node": ">=18"
   },
   "dependencies": {
-    "address-rfc2821": "^2.1.1",
-    "address-rfc2822": "^2.1.0",
+    "address-rfc2821": "^2.1.2",
+    "address-rfc2822": "^2.2.0",
     "async": "^3.2.5",
     "daemon": "~1.1.0",
     "ipaddr.js": "~2.1.0",

--- a/tests/outbound/index.js
+++ b/tests/outbound/index.js
@@ -166,10 +166,7 @@ exports.get_tls_options = {
         const testDir = path.resolve('tests');
         this.outbound.config = this.outbound.config.module_config(testDir);
         this.obtls.test_config(tls_socket.config.module_config(testDir), this.outbound.config);
-        this.obtls.init(() => {
-            done();
-        })
-
+        this.obtls.init(done)
     },
     tearDown: done => {
         delete process.env.HARAKA_TEST_DIR;

--- a/tests/outbound/mx_lookup.js
+++ b/tests/outbound/mx_lookup.js
@@ -1,0 +1,45 @@
+
+const mx = require('../../outbound/mx_lookup');
+
+exports.lookup_mx = {
+    'MX records for example.com (await)': async test => {
+        test.expect(1)
+        const r = await mx.lookup_mx('example.com');
+        test.deepEqual(r, [ { exchange: '', priority: 0 }])
+        test.done()
+    },
+    'MX records for example.com (cb)': test => {
+        test.expect(1)
+        mx.lookup_mx('example.com', (err, r) => {
+            test.deepEqual(r, [ { exchange: '', priority: 0 }])
+            test.done()
+        });
+    },
+    'MX records for tnpi.net (await)': async test => {
+        test.expect(1)
+        const r = await mx.lookup_mx('tnpi.net');
+        test.deepEqual(r, [ { exchange: 'mail.theartfarm.com', priority: 10 }])
+        test.done()
+    },
+    'MX records for tnpi.net (cb)': test => {
+        test.expect(1)
+        mx.lookup_mx('tnpi.net', (err, r) => {
+            test.deepEqual(r, [ { exchange: 'mail.theartfarm.com', priority: 10 }])
+            test.done()
+        });
+    },
+    'MX records for gmail.com (await)': async test => {
+        test.expect(1)
+        const r = await mx.lookup_mx('gmail.com');
+        // console.log(r)
+        test.ok(r.length)
+        test.done()
+    },
+    'MX records for gmail.com (callback)': test => {
+        test.expect(1)
+        mx.lookup_mx('gmail.com', (err, r) => {
+            test.ok(r.length)
+            test.done()
+        });
+    },
+}

